Maintenance locks and interactivity

derrickstolee opened the pull request:
* [x] This is an early version of work ~~already under review upstream~~.
* [x] This change only applies to interactions with Azure DevOps and the
      GVFS Protocol.
* [ ] This change only applies to the virtualization hook and VFS for Git.

This set of patches has a little of both:

1. We need to revert a custom change around `maintenance.lock` files that was only in `microsoft/git`. The "fix" actually caused a worse situation where many background maintenance processes would pile up.
2. Create the `credential.interactive` config as an official Git config and use it in Git to prevent any chance of a foreground username/password request. (This is borrowed from GCM, which already respects this.)
3. To help avoid problems, especially when blocked by credentials, add configuration to the background schedule to avoid interactive prompts.
4. To make sure that the new config options are set up in the background schedule, update `scalar reconfigure` to execute `git maintenance start`.

Points 2-4 could maybe to upstream, but we shouldn't wait for that.

I've tested these features locally on Linux and Windows and double-checked that they will prevent the backup of maintenance processes when the credentials become invalid.

derrickstolee committed:
fixup! maintenance: delete stale lock files

This change from microsoft/git#468 is causing multiple maintenance
processes to get blocked on credentials instead of only one. The change
did more harm than good.

This reverts commit 95ed7f6a2b7b99dfdb1910e8ef7ed85207c27e61.

derrickstolee committed:
maintenance: add custom config to background jobs

At the moment, some background jobs are getting blocked on credentials
during the 'prefetch' task. This leads to other tasks, such as
incremental repacks, getting blocked. Further, if a user manages to fix
their credentials, then they still need to cancel the background process
before their background maintenance can continue working.

Update the background schedules for our four scheduler integrations to
include these config options via '-c' options:

 * 'credential.interactive=false' will stop Git and some credential
   helpers from prompting in the UI (assuming the '-c' parameters are
   carried through and respected by GCM).

 * 'core.askPass=true' will replace the text fallback for a username
   and password into the 'true' command, which will return a success in
   its exit code, but Git will treat the empty string returned as an
   invalid password and move on.

We can do some testing that the credentials are passed, at least in the
systemd case due to writing the service files.

Signed-off-by: Derrick Stolee <derrickstolee@github.com>

jeffhostetler commented on builtin/gc.c:
```diff
@@ -1678,6 +1657,9 @@ static const char *get_frequency(enum schedule_priority schedule)
 
+static const char *custom_background_config =
+	"-c credential.interactive=false -c cred.askPass=false";
+
```
Just a nit.  As an alternative to defining a static string here, you could #define a macro and use that in each of the fprintf() format strings below (like we do for the `PRIuMAX` or `PRItime` strings).

jeffhostetler commented on credential.c:
```diff
@@ -301,6 +302,14 @@ static int run_credential_helper(struct credential *c,
+
+	if (!git_config_get_int("credential.timeout", &config_seconds) &&
+	    config_seconds >= 0)
+		select_timeout.tv_sec = config_seconds;
```
Am I reading this right?  If `credential.timeout` is not defined in the config, we'll default to 5 minutes?  So to avoid the timeout and get the legacy wait, we need to have `credential.timeout=0` in the config.  Not questioning, just confirming intent.

derrickstolee replied:
Yes, that's the intent.

mjcheetham replied:
Just a note for GCM.. this is a setting we should ignore ourselves :)
As in, GCM itself shouldn't have a timeout, since Git is the one doing the timeout and kill.

jeffhostetler commented on credential.c:
```diff
@@ -323,6 +332,20 @@ static int run_credential_helper(struct credential *c,
+				return -1;
+			}
+		}
+
```
I'm assuming that if we get a single byte from GCM, we'll get the full response so we don't need to do something more fancy.  GCM will either be stuck or not before the first byte is returned.

derrickstolee replied:
That's also my expectation. The response is not provided until after the "blocking" part of GCM is complete. @mjcheetham or @ldennington might be able to confirm for sure.

mjcheetham replied:
GCM only writes back to `stdout` once it has the complete result. If we've written anything back, we will soon exit.

jeffhostetler commented on credential.c:
```diff
@@ -323,6 +332,20 @@ static int run_credential_helper(struct credential *c,
+
+			if (!FD_ISSET(helper.out, &readfds)) {
+				/* Timeout complete before helper.out has bytes to read. */
+				kill_child_command(&helper);
```
This bothers me a bit.  I'm wondering if we should just insert a `kill()` here and then let the existing call to `finish_command()` at the bottom wait for the child.  The existing finish code will end up calling `wait_or_whine()` which normalizes the child's exit code, so we'll be consistent.  And then it'll emit
the `child_exit` and other cleanup.

You could also add a trace2_data after the `kill()` to log that we sent the signal, but that is not strictly necessary.

jeffhostetler replied:
Also, then you won't need the `kill_child_command()` function at all.  That might make upstream integration easier.

derrickstolee replied:
Interesting. I hadn't considered that, because I expected the behavior to be "kill the child then leave the method" and not to go to a later `finish_command()`. I will play with this when I have a better testing environment ready.

mjcheetham commented on builtin/gc.c:
```diff
@@ -1678,6 +1657,9 @@ static const char *get_frequency(enum schedule_priority schedule)
 }
 
+static const char *custom_background_config =
+	"-c credential.interactive=false -c cred.askPass=false";
```
I wonder if it's worth trying to upstream the concept of `credential.interactive`, and have it mean "helpers should not interact with the user".

Right now, this is a GCM specific thing.

Alternatively, perhaps a `core.interactive` or `core.background` setting would make more sense? To indicate that "this instance of Git is running non-interactively or in the background", and that is a signal that helpers can pick up on to mean "no prompt"?

jeffhostetler replied:
It might be simpler to have `core.background` as a single value.  Then GCM or the ask-pass or `whatever` code could do the right thing without us having to enumerate the various flags here.

jeffhostetler commented:
Does `gc` know enough about it's environment to know whether to set the background bit?  That is, if `gc` is run by `maintenance`, then yes, but if `gc` is run by a foreground `fetch`, then no.  Right??

jeffhostetler commented:
Same question about `credential.c` WRT the timeout.  If I interactively do a push/fetch and get stuck behind a cred prompt -- and have to dig my phone out of my backpack or while I go get coffee, should it always timeout/abort?  Or should the timeout only be enabled in the maintenance case?

Do we still need the timeout if GCM respects the suggested `core.background` flag ??

derrickstolee committed:
scalar: configure maintenance during 'reconfigure'

The 'scalar reconfigure' command is intended to update registered repos
with the latest settings available. However, up to now we were not
reregistering the repos with background maintenance.

In particular, this meant that the background maintenance schedule would
not be updated if there are improvements between versions.

Be sure to register repos for maintenance during the reconfigure step.

Signed-off-by: Derrick Stolee <derrickstolee@github.com>

derrickstolee commented:
> Same question about `credential.c` WRT the timeout. If I interactively do a push/fetch and get stuck behind a cred prompt -- and have to dig my phone out of my backpack or while I go get coffee, should it always timeout/abort? Or should the timeout only be enabled in the maintenance case?

I could consider leaving it as a non-timeout case for foreground and set the config in the maintenance scheduler (like we are already doing for the interactivity bit).

> Do we still need the timeout if GCM respects the suggested `core.background` flag ??

Is that a thing? One thing I _think_ happens is that background jobs don't have a TTY, so we won't get blocked on Git's request for a username/password (which during my local testing requires setting `GIT_TERMINAL_PROMPT=0`).

jeffhostetler commented:
> > Same question about `credential.c` WRT the timeout. If I interactively do a push/fetch and get stuck behind a cred prompt -- and have to dig my phone out of my backpack or while I go get coffee, should it always timeout/abort? Or should the timeout only be enabled in the maintenance case?
> 
> I could consider leaving it as a non-timeout case for foreground and set the config in the maintenance scheduler (like we are already doing for the interactivity bit).

I'll wait for @mjcheetham opinion here, but I'm wondering if we want to change the foreground behavior here.

> 
> > Do we still need the timeout if GCM respects the suggested `core.background` flag ??
> 
> Is that a thing? One thing I _think_ happens is that background jobs don't have a TTY, so we won't get blocked on Git's request for a username/password (which during my local testing requires setting `GIT_TERMINAL_PROMPT=0`).

I'm not sure. There are too many child processes with STDIN/OUT bound to a pipe from the parent process for me to casually trust `isatty(fd)` ...

derrickstolee commented:
> > > Do we still need the timeout if GCM respects the suggested `core.background` flag ??
> > 
> > 
> > Is that a thing? One thing I _think_ happens is that background jobs don't have a TTY, so we won't get blocked on Git's request for a username/password (which during my local testing requires setting `GIT_TERMINAL_PROMPT=0`).
> 
> I'm not sure. There are too many child processes with STDIN/OUT bound to a pipe from the parent process for me to casually trust `isatty(fd)` ...

What's even worse is that the Git prompt for username/password goes through `git_terminal_prompt()` which pulls the terminal directly from the environment, and ignores something like redirecting `/dev/null` into `stdin`.

derrickstolee commented:
End-to-End Testing Report
-------------------------

After some local testing in Linux helped identify some issues, I generated [a Windows installer](https://github.com/microsoft/git/releases/download/untagged-0394b0835d43626982ba/Git-2.41.0.vfs.0.99-64-bit.exe) and ran it on my Windows machine. Along with an earlier version of the installer, I was able to find out this information:

* I had made a mistake of using the wrong config key name sometimes when I meant `core.askpass`. Further, using "`false`" results in some failures that are not helpful. Instead, using `core.askpass=echo` makes Git skip the feature and move on completely.
* In addition to the `core.askpass` config key, there is a `GIT_ASKPASS` environment variable that overrides the config key. It's set by VS Code (at least in Remote SSH connections like I use).
* Even when using an alternative `core.askpass`, Git has an interactive username/password request that goes through the terminal. See `git_terminal_prompt()` down deep below `credential_getpass()` for this info. This doesn't block when using a background job, since there is no terminal, but it makes testing challenging. 
* The previous two bullets might be good reasons to introduce `credential.interactive` upstream: we can block this behavior when the user is requesting it.
* While the `select()` method works appropriately on Linux, on Windows it seems to be returning immediately as "the readfd is ready" instead of waiting for a byte of data to be sent. Because this isn't helping, I will most likely remove the timeout feature from this pull request.
* _However_ I was able to get things to work by using the `-c credential.interactive=false -c core.askpass=echo` options. These allow fetches to work when the credentials are valid, and the fetches complete (with expected failure) when the credentials are not valid. This will solve our problems with background fetches getting blocked on credentials.
* In order to get these custom configs into the background jobs, we actually need the schedule to be updated during upgrade. I added a patch to make `scalar reconfigure` run `git maintenance start` in the necessary repos, giving us an updated schedule.

I need to update this branch with the full learnings here, and add some tests now that we understand the code necessary to get the features we need.

derrickstolee committed:
credential: add new interactive config option

When scripts or background maintenance wish to perform HTTP(S) requests,
there is a risk that our stored credentials might be invalid. At the
moment, this causes the credential helper to ping the user and block the
process. Even if the credential helper does not ping the user, Git falls
back to the 'askpass' method, which includes a direct ping to the user
via the terminal.

Even setting the 'core.askPass' config as something like 'echo' will
causes Git to fallback to a terminal prompt. It uses
git_terminal_prompt(), which finds the terminal from the environment and
ignores whether stdin has been redirected. This can also block the
process awaiting input.

Create a new config option to prevent user interaction, favoring a
failure to a blocked process.

The chosen name, 'credential.interactive', is taken from the config
option used by Git Credential Manager to already avoid user
interactivity, so there is already one credential helper that integrates
with this option. However, older versions of Git Credential Manager also
accepted other string values, including 'auto', 'never', and 'always'.
The modern use is to use a boolean value, but we should still be
careful that some users could have these non-booleans. Further, we
should respect 'never' the same as 'false'. This is respected by the
implementation and test, but not mentioned in the documentation.

The implementation for the Git interactions takes place within
credential_getpass(). The method prototype is modified to return an
'int' instead of 'void'. This allows us to detect that no attempt was
made to fill the given credential, changing the single caller slightly.

Also, a new trace2 region is added around the interactive portion of the
credential request. This provides a way to measure the amount of time
spent in that region for commands that _are_ interactive. It also makes
a conventient way to test that the config option works with
'test_region'.

Signed-off-by: Derrick Stolee <derrickstolee@github.com>

derrickstolee committed:
fixup! scalar: do initialize `gvfs.sharedCache`

In this commit, we added the 'credential.interactive=never' option to
unattended scalar options. This should be changed to 'false' to match
the modern use of this config option.

But also, we have a test that requires using askpass to get credentials,
but the test is in unattended mode. Fix that test to include
'credential.interactive=true' to bypass this issue.

jeffhostetler commented:
nit: typo in commit message of "add new interactive config option": `carefult`

jeffhostetler commented:
i didn't see anything else.  thanks!

dscho commented on credential.c:
```diff
@@ -196,14 +198,30 @@ static char *credential_ask_one(const char *what, struct credential *c,
+	char *value;
+	if (!git_config_get_maybe_bool("credential.interactive", &interactive) &&
+	    !interactive)
+		return -1;
```
Do we want to trace the fact that the interactive credential prompt was skipped?

derrickstolee replied:
Interesting idea. I was thinking that the lack of the other region would be enough.

Do you propose using a `trace2_printf()`? or what kind of indicator? I'm not familiar with an example of this kind of tracing.

jeffhostetler replied:
you could do something like a trace2_data_intmax(... "credential", "interactive/skipped", 1)

only log the true case.

dscho commented on t/t5551-http-fetch-smart.sh:
```diff
@@ -186,6 +186,28 @@ test_expect_success 'clone from password-protected repository' '
+test_expect_success 'credential.interactive=false skips askpass' '
+	set_askpass bogus nonsense &&
+	(
+		GIT_TRACE2_EVENT="$(pwd)/interactive-true" &&
+		export GIT_TRACE2_EVENT &&
+		test_must_fail git clone --bare "$HTTPD_URL/auth/smart/repo.git" interactive-true-dir &&
+		test_region credential interactive interactive-true &&
```
I have a slight preference towards avoiding a subshell, but won't insist on it. That is, this could be written as:

```sh
test_must_fail env GIT_TRACE2_EVENT="$(pwd)/interactive-true" \
git clone --bare "$HTTPD_URL/auth/smart/repo.git" interactive-true-dir &&
test_region credential interactive interactive-true &&
```

derrickstolee replied:
Ah, the `env` trick is what I was forgetting. Thanks.

derrickstolee replied:
`test_must_fail: only 'git' is allowed: ` 😞

dscho replied:
This works for me:

```diff
diff --git a/t/t5551-http-fetch-smart.sh b/t/t5551-http-fetch-smart.sh
index 6e70bca95a4..d56e3820cb0 100755
--- a/t/t5551-http-fetch-smart.sh
+++ b/t/t5551-http-fetch-smart.sh
@@ -188,24 +188,19 @@ test_expect_success 'clone from password-protected repository' '
 
 test_expect_success 'credential.interactive=false skips askpass' '
 	set_askpass bogus nonsense &&
-	(
-		GIT_TRACE2_EVENT="$(pwd)/interactive-true" &&
-		export GIT_TRACE2_EVENT &&
-		test_must_fail git clone --bare "$HTTPD_URL/auth/smart/repo.git" interactive-true-dir &&
-		test_region credential interactive interactive-true &&
-
-		GIT_TRACE2_EVENT="$(pwd)/interactive-false" &&
-		export GIT_TRACE2_EVENT &&
-		test_must_fail git -c credential.interactive=false \
+	test_must_fail env GIT_TRACE2_EVENT="$(pwd)/interactive-true" \
+		git clone --bare "$HTTPD_URL/auth/smart/repo.git" interactive-true-dir &&
+	test_region credential interactive interactive-true &&
+
+	test_must_fail env GIT_TRACE2_EVENT="$(pwd)/interactive-false" \
+		git -c credential.interactive=false \
 			clone --bare "$HTTPD_URL/auth/smart/repo.git" interactive-false-dir &&
-		test_region ! credential interactive interactive-false &&
+	test_region ! credential interactive interactive-false &&
 
-		GIT_TRACE2_EVENT="$(pwd)/interactive-never" &&
-		export GIT_TRACE2_EVENT &&
-		test_must_fail git -c credential.interactive=never \
+	test_must_fail env GIT_TRACE2_EVENT="$(pwd)/interactive-never" \
+		git -c credential.interactive=never \
 			clone --bare "$HTTPD_URL/auth/smart/repo.git" interactive-never-dir &&
-		test_region ! credential interactive interactive-never
-	)
+	test_region ! credential interactive interactive-never
 '
 
 test_expect_success 'clone from auth-only-for-push repository' '
```

dscho commented on builtin/gc.c:
```diff
@@ -1657,6 +1657,10 @@ static const char *get_frequency(enum schedule_priority schedule)
 
+#define CUSTOM_BACKGROUND_CONFIG \
+	"-c credential.interactive=false " \
+	"-c core.askPass=echo "  /* 'echo' returns empty string. */
```
According to https://git-scm.com/docs/git-config#Documentation/git-config.txt-coreaskPass, the `askPass` program "shall be given a suitable prompt as command-line argument and write the password on its STDOUT."

Therefore I suspect that the `echo` command will not actually return an empty string, but instead the prompt it was given.

I further suspect that this prompt will then be used as password and a network roundtrip will result in a failed authentication.

In other instances where a program needs to be called that basically does nothing but succeed, Git seems to use `:` instead (see e.g. https://github.com/git/git/blob/v2.42.0/editor.c#L64). Would this work here, too?

derrickstolee replied:
That's a good point. The extra round trip can probably be avoided using `true` and Git will figure out that a password was not provided.

dscho commented on builtin/gc.c:
```diff
@@ -1863,6 +1867,10 @@ static int launchctl_schedule_plist(const char *exec_path, enum schedule_priorit
+		   "<string>-c</string>\n"
+		   "<string>credential.interactive=false</string>\n"
+		   "<string>-c</string>\n"
+		   "<string>core.askPass=echo</string>\n" /* 'echo' returns empty string. */
```
There is a slight concern that this is supposed to be equivalent to `CUSTOM_BACKGROUND_CONFIG` but not derived from it, and therefore might go stale.

It might be overkill, but this here would work, I think:

```c
struct strbuf extra_config = STRBUF_INIT;
const char *p = CUSTOM_BACKGROUND_CONFIG;

while (*p) {
  const char *space = strchrnul(p, ' ');

  strbuf_addf(&extra_config, "<string>%.*s</string>\n", (int)(space - p), p);

  p = *space ? space + 1 : space;
}
```

and then use it as another `%s` in the `preamble`.

_Alternatively_, and much more elegant, define this instead:

```c
#define CUSTOM_BACKGROUND_CONFIG_1 "credential.interactive=false"
#define CUSTOM_BACKGROUND_CONFIG_2 "core.askPass=echo"  /* 'echo' returns empty string. */
```

and then use those constants in ` launchctl_schedule_plist()`, `schtasks_schedule_task()`, `crontab_update_schedule()` and in `systemd_timer_write_service_template()`.

derrickstolee replied:
Here's my attempt at keeping things working (also, replacing `echo` with `true`):

```c
const char *extraconfig[] = {
	"credential.interactive=false",
	"core.askPass=true", /* 'true' returns success, but no output. */
	NULL
};

static const char *get_extra_config_parameters(void) {
	static const char *result = NULL;
	struct strbuf builder = STRBUF_INIT;

	if (result)
		return result;

	for (const char **s = extraconfig; s && *s; s++)
		strbuf_addf(&builder, "-c %s ", *s);

	result = strbuf_detach(&builder, NULL);
	return result;
}

static const char *get_extra_launchctl_strings(void) {
	static const char *result = NULL;
	struct strbuf builder = STRBUF_INIT;

	if (result)
		return result;

	for (const char **s = extraconfig; s && *s; s++) {
		strbuf_addstr(&builder, "<string>-c</string>\n");
		strbuf_addf(&builder, "<string>%s</string>\n", *s);
	}

	result = strbuf_detach(&builder, NULL);
	return result;
}
```

dscho commented on t/t9210-scalar.sh:
```diff
@@ -176,8 +176,9 @@ test_expect_success 'scalar reconfigure' '
-	test true = "$(git -C one/src config core.preloadIndex)"
+	GIT_TRACE2_EVENT="$(pwd)/reconfigure" scalar reconfigure -a &&
+	test true = "$(git -C one/src config core.preloadIndex)" &&
+	test_subcommand git maintenance start <reconfigure
```
Do we maybe want to test specifically that the maintenance was re-scheduled? I.e. this instead?

```diff
diff --git a/t/t9210-scalar.sh b/t/t9210-scalar.sh
index 4432a30d10b..83c87ab744e 100755
--- a/t/t9210-scalar.sh
+++ b/t/t9210-scalar.sh
@@ -169,7 +169,9 @@ test_expect_success 'scalar reconfigure' '
 	scalar reconfigure one &&
 	test true = "$(git -C one/src config core.preloadIndex)" &&
 	git -C one/src config core.preloadIndex false &&
+	rm one/src/cron.txt &&
 	scalar reconfigure -a &&
+	test_is_file one/src/cron.txt &&
 	test true = "$(git -C one/src config core.preloadIndex)"
 '
 
```

dscho reviewed (commented):
I am optimistic that this will address the reported problems.

A couple of feedback comments about the commit messages:
- In the second commit's message: "caues" -> "cause", "modifed" -> "modified".
- The third commit's message mentions "GCM" without prior explanation of the acronym; I would recommend using "Git Credential Manager" here instead.
